proto/substrait/algebra.proto

```diff
@@ -477,6 +481,60 @@ message WriteRel {
   }
 }
 
+// The hash equijoin join operator will build a hash table out of the right input based on a set of join keys.
+// It will then probe that hash table for incoming inputs, finding matches.
+message HashJoinRel {
+  RelCommon common = 1;
+  Rel left = 2;
+  Rel right = 3;
+  Expression expression = 4;
```

Comment:
I think we should specialize these fields for what a dumb consumer would want. I think that is probably:

FieldReference[] left_keys (hash table)
FieldReference[] left_keys
Expression during_join_filter (join conditions)
Expression post_join_filter (afterwards)

We should also validate that the join types are valid in this situation.

Reply:
I like this idea.  @jvanstraten @vibhatha and I have struggled at times with the mismatch that Acero expects simple field references (except in the project relation) in many places the official spec contains full expressions.  This comment helped me crystallize that this is because those spots with full expressions are simply more logical than physical.

Reply:
+1; I don't personally know what it should look like, but for physical relations I wasn't expecting expressions at all. I'd go as far as to say that the and-tree-of-equalities expression should be entirely reduced to just a bunch of FieldReferences, so a consumer doesn't have to match and unpack something involving something as complex as extension functions just to understand what the producer meant.

Reply:
> the and-tree-of-equalities expression should be entirely reduced to just a bunch of FieldReferences

For reference, in Acero, we currently only support a vector of field references (one from the left and one from the right) and, for each field, an enum of `equal` or `is_not_distinct_from`.  Though that would limit this to equi-joins.

Reply:
Reading  [velox ](https://github.com/facebookincubator/velox)code,  hash join and merge join also support a vector of field references for 'lefyKeys' and 'rightKeys', but how can I pass the [outputType ](https://github.com/facebookincubator/velox/blob/main/velox/core/PlanNode.h#L1086)from proto , how can I pass it from proto ?

Reply:
I think it is valuable to support the equal/is not distinct concepts for each of the key pairings here. Probably makes most sense to use a message similar to what it sounds like acero does (e.g. left_key, right_key, equality_condition)

WRT to join type, I'm inclined to only introduce the join types people are using for each of these. We can introduce more as people have them. As such, I'd suggest we have separate enums for each (HashJoinType, MergeJoinType, etc). What do others think?

Reply:
> As such, I'd suggest we have separate enums for each (HashJoinType, MergeJoinType, etc). What do others think?

I agree with the thought because they seem different enough (and also agree with your other paragraph). However, I'd personally opt against adding new enum types in the protos, because enums in general are pretty fragile compared to everything else in protobuf (lack of proper scoping, annoyingly long names in JSON, representing unknown values with invalid enum values in C, easy for consumers to forget to handle unrecognized values, can't add options that contain values later like you can with e.g. Rust enums, etc). A message containing a `oneof`s with (initially) empty messages seems superior to me in every way. I don't think it even costs more in the binary serialization.

Reply:
I'm actually pretty supportive of this pattern generally. It feels much more like rust enums.